### PR TITLE
Make nodes callback-based rather than polling

### LINKED ISSUE #227: Make ROS nodes interrupt-based
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
Our ROS nodes should only run when the callbacks are triggered, rather than poll all the time.
* The `network_input` should only process and publish data when it is received from the network
* The `AI` node should only make a new decision when it receives new data
* The `grsim_communication` node should only send new commands when it receives new field data or new primitives
* Same as above for `robot_communication`

**Describe the solution you'd like**
Rather than have a `main` function with an infinite loop, move the logic to the callback functions and call `ros::spin()` rather than `ros::spinOnce()` in the loop.

**Describe alternatives you've considered**
Leaving things as they are. This should be a better solution that makes managing timestamps easier (no need to update them without new data), and reduce the load on people's CPUs (as per #153 )

**Additional context**
None

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Made nodes run when callbacks are triggered rather than polling all the time.

`ai_logic`, `radio_communication`, and `grsim_communication` nodes now only run when they receive a new `World` message from the ROS callbacks. The `network_input` node has been rewritten to only filter and publish data when it is received from the socket, without polling for it.

TODO: Make sure the `network_input` node is able to shut down properly

### Testing Done

Manually verified nodes still run and that we can perform basic operations like sending a primitive to grsim.

CPU STATS
Before:
* ai_logic: 175% (WTF!?)
* network_input: 102%
* grsim_communication: 50%

After:
* network_input: 1%
* ai_logic: 1%
* grsim_communication: 1%
* CPU usage _looks_ way lower than before

### Resolved Issues

resolves #227 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
@qgolsteyn please try this out and let me know if you notice improvement on your (weak) machine.

File: src/thunderbots/software/ai/ai.cpp [hunk 1]
```diff
@@ -1,23 +1,14 @@
...
-    : world(world),
-      navigator(std::make_unique<RRTNav>()),
-      high_level(std::make_unique<STP_HL>())
+AI::AI() : navigator(std::make_unique<RRTNav>()), high_level(std::make_unique<STP_HL>())
 {
 }
 
-std::vector<std::unique_ptr<Primitive>> AI::getPrimitives(
-    const Timestamp &timestamp) const
+std::vector<std::unique_ptr<Primitive>> AI::getPrimitives(const World &world) const
```
Comment (garethellis0) [inline src/thunderbots/software/ai/ai.cpp:10]: I like this :+1:

File: src/thunderbots/software/ai/main.cpp [hunk 2]
```diff
@@ -19,58 +15,32 @@
...
-    thunderbots_msgs::Team enemy_team_msg = *msg;
-
-    Team enemy_team = Util::ROSMessages::createTeamFromROSMessage(enemy_team_msg);
+    // Put these Primitives into a message and publish it
+    thunderbots_msgs::PrimitiveArray primitive_array_message;
+    for (auto const &prim : assignedPrimitives)
+    {
+        primitive_array_message.primitives.emplace_back(prim->createMsg());
+    }
+    primitive_publisher.publish(primitive_array_message);
```
Comment (garethellis0) [inline src/thunderbots/software/ai/main.cpp:40]: _much_ cleaner

File: src/thunderbots/software/radio_communication/main.cpp [hunk 3]
```diff
@@ -1,61 +1,37 @@
...
-}  // namespace
-// Callbacks
 void primitiveUpdateCallback(const thunderbots_msgs::PrimitiveArray::ConstPtr& msg)
 {
+    std::vector<std::unique_ptr<Primitive>> primitives;
     thunderbots_msgs::PrimitiveArray prim_array_msg = *msg;
     for (const thunderbots_msgs::Primitive& prim_msg : prim_array_msg.primitives)
     {
         primitives.emplace_back(AI::Primitive::createPrimitiveFromROSMessage(prim_msg));
     }
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/main.cpp]: Should we be doing something else with these primitives?

File: src/thunderbots/software/radio_communication/main.cpp [hunk 4]
```diff
@@ -1,61 +1,37 @@
...
-    while (ros::ok())
-    {
-        // Clear all primitives each tick
-        primitives.clear();
-
-        // Spin once to let all necessary callbacks run
-        // The callbacks will populate the primitives vector
-        ros::spinOnce();
-    }
+    ros::spin();
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/main.cpp:83]: Quick comment on what `ros::spin()` does would be good (same goes for other instances of it in this PR)

File: src/thunderbots/software/util/constants.h [hunk 5]
```diff
@@ -28,6 +28,11 @@ namespace Util
         static const std::string SSL_VISION_MULTICAST_ADDRESS = "224.5.23.2";
         static const unsigned short SSL_VISION_MULTICAST_PORT = 10020;
 
+        // GrSim networking and communication
+        // TODO: BETTER NAMES
```
Comment (garethellis0) [inline src/thunderbots/software/util/constants.h:32]: TODO?

File: src/thunderbots/software/util/ros_messages.cpp [hunk 6]
```diff
@@ -165,5 +165,20 @@ namespace Util
...
+        World createWorldFromROSMessage(const thunderbots_msgs::World& world_msg)
+        {
+            Ball ball          = createBallFromROSMessage(world_msg.ball);
+            Team friendly_team = createTeamFromROSMessage(world_msg.friendly_team);
+            Team enemy_team    = createTeamFromROSMessage(world_msg.enemy_team);
+            Field field        = createFieldFromROSMessage(world_msg.field);
+
+            // TODO: ???
+            //            RefboxGamState refbox_gamestate =
+            //            createGameStateFromROSMessage(world_msg.refbox_data);
```
Comment (garethellis0) [inline src/thunderbots/software/util/ros_messages.cpp:178]: TODO?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
- sorry, just noticed this was WIP, but pointed out all the TODO's,  ¯\_(ツ)_/¯
- should there be some changes in network input? 

Overall this is looking really good, makes things much easier to follow, which is a great bonus.

File: src/thunderbots/software/radio_communication/main.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/radio_communication/main.cpp]: Yes, but that's up to @henrymliu 's later PRs

File: src/thunderbots/software/radio_communication/main.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/radio_communication/main.cpp:83]: comments added

**Comment by MathewMacDougall** [comment]
I have now added the changes to `network_input` that @garethellis0 predicted. I have updated the description and test results accordingly.

The `network_input` node now only filters/publishes packets when they are received by the socket, without us having to poll.

This has **_significantly_** reduced CPU load at seemingly no performance loss. We are still able to receive/publish data at ~240 HZ which is expected (4 cameras at 60HZ each)

The last thing I need to do is make sure the `io_service` thread and `ros::spin` are playing nice together and shut down properly. Right now the `network_input` hangs on shutdown before ros escalates the termination signal to `SIGTERM`. Other than that, this PR should be good to go, so feel free to review it like it's no longer WIP.

I also encourage you to try this on your machines to confirm the reduced CPU load and make sure things behave reasonably sane still.

File: src/thunderbots/software/radio_communication/main.cpp (hunk 3 above)
Comment (garethellis0) [reply src/thunderbots/software/radio_communication/main.cpp]: Ok, can we leave a TODO with an issue reference here?

File: src/thunderbots/software/network_input/main.cpp [hunk 7]
```diff
@@ -49,17 +31,80 @@ int main(int argc, char** argv)
...
+
+    // Store the state of the most up to date World message
+    thunderbots_msgs::World world_msg;
+
+    // Create the io_service that will be used to service all network requests
+    boost::asio::io_service io_service;
+
+    // The function that will be run on every received SSL_WrapperPacket
+    auto filter_and_publish_vision_data = [&backend, world_publisher,
+                                           &world_msg](SSL_WrapperPacket packet) {
```
Comment (garethellis0) [inline src/thunderbots/software/network_input/main.cpp]: This should not be a lambda function. It's _way_ too big.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
I think we should probably have another class that encompasses all the functionality currently in `network_input/main.cpp`, there's a lot there that would benefit from some organization.

The issue with you segfaulting when you exit is probably because you're exiting without waiting for the thread to finish. When the destructor for a live thread object is called, it calls `std::terminate`, which kills everything. You need to somehow signal to your thread that it's time to end (probably via a mutex-protected bool) and then `join` the thread to let it exit. See the destructor for the `PassGenerator` class for an example of this. This will likely be easier/cleaner when all this functionality is in a class.

Overall I really like this though! Confirmed `network_input` taking 25% of my cpu, grsim taking 17%

File: src/thunderbots/software/radio_communication/main.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/radio_communication/main.cpp]: Henry's PRs are merged and I've rebased on top of them so this is no longer an issue.

File: src/thunderbots/software/network_input/main.cpp (hunk 7 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/main.cpp]: Refactored to wrap the clients and lambda functions in a general `NetworkClient`

File: src/thunderbots/software/network_input/networking/network_client.h [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    explicit NetworkClient(ros::NodeHandle& node_handle);
+
+    /**
+     * Safely destructs this NetworkClient object. Stops any running IO services and
+     * gracefully joins the thread before exiting.
+     */
+    ~NetworkClient();
+
+    // Delete the copy and assignment operators because this class really shouldn't need
+    // them and we don't want to risk doing anything nasty with the internal thready this
```
Comment (garethellis0) [inline src/thunderbots/software/network_input/networking/network_client.h]:
``` suggestion
    // them and we don't want to risk doing anything nasty with the internal threading this
````

File: src/thunderbots/software/network_input/networking/network_client.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    // Set up publishers
+    world_publisher = node_handle.advertise<thunderbots_msgs::World>(
+        Util::Constants::NETWORK_INPUT_WORLD_TOPIC, 1);
+    gamecontroller_publisher = node_handle.advertise<thunderbots_msgs::RefboxData>(
+        Util::Constants::NETWORK_INPUT_GAMECONTROLLER_TOPIC, 1);
+
+    // Set up our connection over udp to receive vision packets
+    // NOTE: We do this before initializing the ROS node so that if it
+    // fails because there is another instance of this node running
+    // and connected to the port we want, we don't kill that other node.
```
Comment (garethellis0) [inline src/thunderbots/software/network_input/networking/network_client.cpp]: Do we? Where are we initializing the ROS stuff?

**Comment by garethellis0** [comment]
This looks good, just two questions above.

**Review by garethellis0 — DISMISSED** [review]

File: src/thunderbots/software/network_input/networking/network_client.cpp (hunk 9 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/networking/network_client.cpp]: Ah this isn't applicable anymore. 

ROS nodes will shut down if a new node is registered with the same name. We need to initialize the ROS node to get the NodeHandle before creating the NetworkClient anyway, so we can't attempt to bind to the network port before registering the node. This isn't that important anyway.

Comments removed.

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

**Review by garethellis0 — APPROVED** [review]

**Pull request merged**